propagate EOF from stdin to socket, from socket to stdout; rewrite listen mode

lersek opened the pull request:
This patch series fixes https://github.com/u-root/u-root/issues/3329.

In order to do that, I ended up rewriting listen mode.

The first 12 patches lay the foundation; they advance in small steps, and their effect is unobservable until later.

Patch 14 fixes https://github.com/u-root/u-root/issues/3329 for connect mode. Because the fix causes the recently added `netstat` integration test (e7b6bdb7129c) to trigger a redirection bug in [gosh](https://github.com/mvdan/sh), patch 13 first updates said `netstat` integration test to work around the gosh redir bug.

Patch 16 is the large one; it fixes https://github.com/u-root/u-root/issues/3329 for listen mode. As you can tell from the first 12 patches in the series, I strongly believe that patch sets should be constructed from small, isolated, dependency-ordered stages. Unfortunately, I couldn't find any way to fix all the issues with listen mode gradually; I opted for a rewrite. That's in patch 16.

For reviewing a rewrite, it doesn't help if `git show` mixes old hunks with tne new. Therefore patch 15 first moves the current listen mode impl to a separate file, temporarily, so that patch 16 can then present the rewrite as new code.

At every stage, the patch series passes:
- `gofmt`, `go vet`, and `go test` in both `pkg/netcat` and `cmds/core/netcat`,
- `go test -v -run TestNetstat` under `integration/generic-tests` (i.e., the `netstat` integration test). (The `netstat` integration test uses `netcat` for setting up a networking scenario that we then query with `netstat`.)
- `golangci-lint run --out-format=line-number --timeout=5m` (using v1.64.5 of golangci-lint)

On top of this patch set, I've been working on a `netcat` integration test, using two VMs that talk to each other. The idea is to put the EOF propagation (fixed here) to use, in various scenarios. Basically, run

~~~
mkfifo fifo
: >fifo &
netcat -l 192.168.0.2 5005 <fifo >fifo &
~~~

in one VM, and run

~~~
seq -w 0 999999 | netcat 192.168.0.2 5005 | shasum
~~~

in the other VM.

The client uploads the `seq` output to the server. The server writes the data to its stdout, and then reads it back (through the FIFO) from its stdin. The server sends the data back to the client via the socket, where we verify the checksum.

This exercises `shutdown(2)` with `SHUT_WR` in both directions, independently, on the socket (`FIN` packets), plus proper closure of stdout (via redirection to `/dev/null`), which is required for producing an EOF condition on the fifo.

Then there are variations of this: the same loopback transfer can be done with client/server roles reversed (i.e., the client too can loop the data back), plus both of those scenarios can be repeated with a unix domain stream socket (using just one VM).

Unfortunately, what breaks this (and makes EOF propagation ultimately untestable) is a [bug in gosh](https://github.com/mvdan/sh/issues/1142). That bug is more or less the same that we work around in this patch set, in patch 13, for the `netstat` integration test. However, the bug cannot be worked around for enabling the EOF propagation test in the `netcat` integration test; the gosh bug needs to be fixed for that goal.

I've been working on a patch for said gosh bug, but after a few hours, it ultimately doesn't look like an implementation bug to me, but a consequence of a design decision (which I disagree with, to the extent that I understand it -- which may not be much, admittedly). Namely, gosh doesn't run the rightmost command of a multi-command pipeline in a subshell, but directly in the parent shell.

Either way, I can't progress with the patch for https://github.com/mvdan/sh/issues/1142 without help, and so I might propose an integration test for `netcat`, later, that does something simpler than looping back data and EOF.

jensdrenhaus commented on cmds/core/netcat/listen.go:
```diff
@@ -108,31 +109,102 @@ func (c *cmd) setupListener(network, address string) (net.Listener, error) {
+		}
+		return
+	}
+	return nil, errors.New("testLimit exhausted")
```
Nit: Actually the error path should be handled in indent code, not in the main path.

lersek replied:
Hmm, I disagree (or at least I don't understand your point).

If `testLimit` is zero (i.e., we are not unit-testing), then the loop condition always evaluates to true, and so the final `return` is never reached. This is intended.

If `testLimit` is nonzero (i.e., we are unit-testing), then the final `return` is reached if and only if we have exhausted the permitted number of `Accept()` attempts.

In the loop body, we have two `return` cases:
- `Accept()` failed, or
- `Accept()` succeeded *and* the peer's address is not forbidden by the allow-list.

If we are unit-testing and we have exhausted the allotted `Accept()` tries, then the loop body shouldn't be entered at all (we shouldn't try further `Accept()`s then).

... I guess what may be a bit non-obvious is the loop condition

~~~
testLimit == 0 || *testID < testLimit
~~~

This is a pattern I frequently use. The imperative programming languages that I know have no "implication operator", that is, no `a --> b`. However, the implication operator is logically equivalent to `!a || b` (they have the same truth tables), and programming languages do have a logical disjunction operator (such as `||`).

In other words, the loop condition expresses, in "pseudo-Go", the following implication (also known as "conditional requirement"):

~~~
	for (testLimit != 0) --> (*testID < testLimit) {
~~~

meaning: "**if** `testLimit` is nonzero, **then** iterate while we have test attempts left, **otherwise** loop forever".

`false` implies anything -- in an implication `a --> b`, if `a` is false, then the whole implication is true (if the condition for the requirement is not met, then the requirement falls away, and the whole implication evalutes to `true`); so the loop will repeat forever.

And then, given that `a --> b` is equivalent to `!a || b`, the pseudo-Go expression `(testLimit != 0) --> (*testID < testLimit)` can be written as the actual Go expression `testLimit == 0 || *testID < testLimit`.

rminnich replied:
it looks ok to me?

jensdrenhaus replied:
With the for loop being the main part of the function - yep I got your point. 
With https://go.dev/wiki/CodeReviewComments#indent-error-flow in mind, it seemed wired on the first look.

jensdrenhaus reviewed (commented):
Very nice refactoring and easy to review due to good patch set design and documentation 💯

The refactoring of the listen function looks way more solid.
I left some inline comments, but nothing serious.

Further, as a general hint:

- Doc comments should start with the name of the type or function, see https://tip.golang.org/doc/comment for reference
- Consider adding build constraints (this could also solve the CI error)

In general, it looks good to me, I need to go over your design behind closing stdout once more.

lersek commented:
> Further, as a general hint:
> 
> * Doc comments should start with the name of the type or function, see https://tip.golang.org/doc/comment for reference

I've now addressed this, for the upcoming v2. It affects patches:
- nr.2 -- pkg/netcat: add TeeWriteCloser type
- nr.5 -- pkg/netcat: add CloseWrite() helper function
- nr.16 -- cmds/core/netcat: rewrite listenForConnections()

> * Consider adding build constraints (this could also solve the CI error)

I've now investigated the CI errors raised for v1. There are five of them, falling in two classes:
- Class 1:
  1. [one error](https://app.circleci.com/pipelines/github/u-root/u-root/8961/workflows/079f8a02-61e1-4ceb-b696-39a71e98f8de/jobs/111063) complains that plan9 does not have the dup2() syscall.
  2. two errors ([1](https://github.com/u-root/u-root/actions/runs/14177106988/job/39714553299?pr=3337), [2](https://github.com/u-root/u-root/actions/runs/14177106988/job/39714553297?pr=3337)) complain that Windows does not have the dup2() syscall.
  3. [another error](https://github.com/u-root/u-root/actions/runs/14177106989/job/39714553420?pr=3337) complains that yet another (unspecified?) environment lacks the dup2() syscall.

  I don't understand these errors. I thought that u-root solely targeted a linux *execution* environment; therefore, relying on linux syscalls -- more precisely, on POSIX APIs -- should be safe. The utilities that u-root builds are composed into a _Linux initial RAM disk_; so why even try to build for the userspace of other operating systems? For example, I don't understand commit eacd20e68aa98 at all (the commit message doesn't explain the idea).

  I can see two approaches, in the tree. One, some commands such as `chroot` are disabled on plan9 and windows altogether. Two, some commands such as `ls` are built for every OS, but platform-specific parts of them are implemented in multiple ways, in separate files, under (e.g.) `pkg/ls/`. (Which seems to imply that u-root tries to **emulate** `ls` on Windows, similarly to, I don't know, Cygwin? What is that good for?)

  I prefer to exclude `netcat`, whole-sale, from any build that does not target a POSIX platform. How can I do that, with the smallest amount of `go:build` restrictions? (In other words, I prefer to follow `chroot`'s example; however, the difference with `chroot` is that `chroot` has only a `cmds/` component, and no `pkg/` component. How can I most simply exclude *both* `pkg/netcat` *and* `cmds/core/netcat` from builds that target a non-POSIX OS?)

- Class 2: the linter [reporting](https://github.com/u-root/u-root/actions/runs/14177107008/job/39714553296?pr=3337) that I should have used param names such as `testID` and `connID`, rather than `testId` and `connId`.

  I have now fixed this (for v2), but it raises further questions:
  - how can I run lint locally? (and... why is linting not a part of `gofmt`, in the first place?)
  - if the linter flags `connId` and `testId`, then why does it **not** flag `myConnId`? (I've renamed the latter too, just in case, but the linter seems inconsequential in any case.)

lersek commented on pkg/netcat/util.go:
```diff
@@ -60,3 +63,44 @@ func (sw *ConcurrentWriter) Write(p []byte) (n int, err error) {
+// "/dev/null" over fd#1.
+type StdoutWriteCloser struct{}
+
+func NewStdoutWriteCloser() *StdoutWriteCloser {
+	return &StdoutWriteCloser{}
+}
```
Thanks -- I've opted for open-coding the `NewStdoutWriteCloser` constructor at the call site. I wouldn't like to make the underlying `io.WriteCloser` parametrizable, as the whole concept is tied to file descriptor 1. (POSIX says that file descriptors 0, 1 and 2 must be valid during the whole lifetime of a process; I paraphrase.)

lersek replied:
> POSIX says that file descriptors 0, 1 and 2 must be valid during the whole lifetime of a process; I paraphrase.

To substantiate:

* [Shell Command Language, 2.9.1.5 Standard File Descriptors](https://pubs.opengroup.org/onlinepubs/9799919799/utilities/V3_chap02.html#tag_19_09_01_05):
  > [...] If a standard utility or a conforming application is executed with file descriptor 0 not open for reading or with file descriptor 1 or 2 not open for writing, the environment in which the utility or application is executed shall be deemed non-conforming, and consequently the utility or application might not behave as described in this standard. [...]
* [`exec`, Description](https://pubs.opengroup.org/onlinepubs/9799919799/functions/exec.html#tag_17_129_03): exact same language as above.
* [`close`, Application Usage](https://pubs.opengroup.org/onlinepubs/9799919799/functions/close.html#tag_17_81_07):
  > [...] Usage of close() on file descriptors STDIN_FILENO, STDOUT_FILENO, or STDERR_FILENO should immediately be followed by an operation to reopen these file descriptors. Unexpected behavior will result if any of these file descriptors is left in a closed state (for example, an [EBADF] error from perror()) or if an unrelated open() or similar call later in the application accidentally allocates a file to one of these well-known file descriptors. Furthermore, a close() followed by a reopen operation (e.g., open(), dup(), etc.) is not atomic; dup2() should be used to change standard file descriptors.

The last citation is echoed by the [Go documentation](https://pkg.go.dev/os#Stderr) too:

> Note that the Go runtime writes to standard error for panics and crashes; closing Stderr may cause those messages to go elsewhere, perhaps to a file opened later.

lersek replied:
I'm not sure about proper etiquette in u-root wrt. closing discussion threads, but I keep a detailed, patch-level changelog between versions of a patch series (I maintain that with `git notes`), and I'll post that in a new comment as soon as I push v2 for review.

Therefore, in order to reduce clutter, I'm going to go ahead, and close those threads *now* that I *claim* to have resolved in v2. I hope that's fine. (In some projects, reviewers like to keep the threads they had opened unresolved until the next version of the series is pushed; then they re-verify, and if they're happy, they resolve those threads themselves. I can work with that too, but it keeps a larger amount of "rolling clutter", so for now (unless told otherwise) I'll go ahead and close these myself.)

lersek commented on pkg/netcat/util.go:
```diff
@@ -60,3 +63,44 @@ func (sw *ConcurrentWriter) Write(p []byte) (n int, err error) {
+		}
+	}
+
+	return syscall.Dup2(int(devNull.Fd()), syscall.Stdout)
```
Yes, thanks -- please see my further questions in the main comment thread.

lersek replied:
The problem is that the desired semantics [*] is impossible to ascertain on platforms that do not conform to POSIX. I can restrict this function to Linux (or  to`!windows && !plan9` anyway), but even for `windows || plan9`, I'd have to provide a stub implementation. And the stub impl might as well call `panic`, or print `sorry, bug`; it will not do what I need it to do. My preference would be to restrict all of `netcat` to `!windows && !plan9`.

[*] The goal is to drop the `netcat` process's reference to the file description that originally underlies its file descriptor 1, all the while keeping the process alive, and file descriptor 1 valid. This is only doable with dup2(). It allows another process referencing the same file description, or (more likely) referencing the same file (inode) through a different file description, to notice netcat's write reference being dropped, while `netcat` continues transfering from stdin to the socket. The reference-dropping is relevant minimally for FIFO's (it allows netcat to produce an EOF for the reader of the FIFO, while netcat continues transfering in the other direction, which might continue indefinitely). Another use case (with netcat's stdout redirected to a regular file) could be the `IN_CLOSE_WRITE` event with `inotify(7)`.

... Well, I guess these use cases aren't defined on non-POSIX anyway, so it shouldn't be a problem if netcat doesn't support them on non-POSIX. Hence, an empty stub should work, after all. I'll look into that.

Still, my general question stands: what sense does it make for u-root to build its commands for anything but Linux? Thank you.

lersek replied:
Using `pkg/ls` as an example, we have:

* `fileinfo_plan9.go` -- `//go:build plan9`
* `fileinfo_tamago.go`-- `//go:build tamago`
* `fileinfo_unix.go` -- `go:build !plan9 && !windows && !tamago`
* `fileinfo_windows.go` -- no build constraint

Why doesn't `fileinfo_windows.go`specify any build constraint? I think it should specify `//go:build windows`. Given that it lacks such a build constraint at the moment, why does it not conflict with, for example, `fileinfo_unix.go`? On Linux, both the `_unix` and `_windows` implementations seem to apply. Why does the build not fail? Thanks.

jensdrenhaus replied:
> My preference would be to restrict all of netcat to !windows && !plan9
Yep, do it, :+1:

lersek replied:
> > My preference would be to restrict all of netcat to !windows && !plan9
> 
> Yep, do it, 👍

[After all](https://github.com/u-root/u-root/pull/3337#discussion_r2023472281), I'd like to try the stub stuff. I hope I can make it work!

lersek replied:
done in v2, to the best of my ability

lersek commented on pkg/netcat/util.go:
```diff
@@ -64,6 +64,20 @@ func (sw *ConcurrentWriter) Write(p []byte) (n int, err error) {
 	return sw.writer.Write(p)
 }
 
+type ConcurrentWriteCloser struct {
+	ConcurrentWriter
+}
+
+func NewConcurrentWriteCloser(wc io.WriteCloser) *ConcurrentWriteCloser {
+	return &ConcurrentWriteCloser{ConcurrentWriter{writer: wc}}
+}
+
+func (swc *ConcurrentWriteCloser) Close() error {
+	swc.ConcurrentWriter.mu.Lock()
+	defer swc.ConcurrentWriter.mu.Unlock()
+	return swc.ConcurrentWriter.writer.(io.WriteCloser).Close()
+}
+
```
I agree, that was my first idea too. However, `ConcurrentWriter` *is* used in itself, in `TestBroadcastMessage` (`core/netcat/listen_test.go`), for unit-testing `Connections.Broadcast` (which indeed does not need the ability to close). Other unit tests (`TestListenForConnections`, `TestWriteFromListenerToConnection`) cover the derived `ConcurrentWriteCloser` struct (`listenForConnections` does require the closer ability).

jensdrenhaus replied:
Okay :+1:

lersek commented on cmds/core/netcat/listen.go:
```diff
@@ -108,31 +109,102 @@ func (c *cmd) setupListener(network, address string) (net.Listener, error) {
+// Accept a connection from "listener" such that the peer address is permitted
+// by the allow list. Drop connections from forbidden peers transparently
+// (apart from counting test attempts during unit testing).
+func (c *cmd) acceptAllowed(listener net.Listener, testLimit uint32, testId *uint32) (conn net.Conn, err error) {
```
I tried to utilize a Go pattern here, intentionally: I'd noticed that named return values -- named components of a return tuple -- could be used as local variables in function bodies, and that this pattern sometimes allowed for more succinct function bodies. And that's the intent in this function, too: we have

~~~
		conn, err = listener.Accept()
		...
		if err != nil {
			return
		}
		...
		return
~~~

I found the ["naked" return statements](https://go.dev/tour/basics/7) very effective; so rather than documentation, they were the goal.

jensdrenhaus replied:
Sure, it is a language feature. Here are some (non canonical) guidelines discussing it:
https://google.github.io/styleguide/go/decisions#named-result-parameters

To me the following is most reasonable:

> [Naked returns](https://tour.golang.org/basics/7) are acceptable only in a small function. Once it’s a medium-sized function, be explicit with your returned values. Similarly, do not name result parameters just because it enables you to use naked returns. [Clarity](https://google.github.io/styleguide/go/guide#clarity) is always more important than saving a few lines in your function.

lersek commented on cmds/core/netcat/listen.go:
```diff
@@ -108,31 +109,102 @@ func (c *cmd) setupListener(network, address string) (net.Listener, error) {
+}
+
 // Connections holds all the active connections of a listener.
 type Connections struct {
```
I agree `Connections` doesn't need to be exported, but here I just stuck with the existent code. I didn't want to make the last patch even larger.

If you think this is important, can we fix it up after this PR, separately? Sticking with `Connections` (upper-case) for now doesn't make anything worse, and renaming it to `connections` within this patch series would cause unpleasant rebase conflicts for me:
- introduce a new patch between patch v1 nr.14 and v1 nr.15, for the rename
- redo patch v1 nr.15 with the rename in place (because the code movement to `listen_conns.go` would now have to refer to the also-unexported, also-renamed constructor `newConnections`)
- redo patch v1 nr.16 with the rename in place (because the new code now has to refer to the new type / constructor / etc).

Alternatively, I could squash the rename into patch 16, but that would make patch 16 even larger.

I prefer to implement this rename (because you *are* right about it) separately, after this PR. The rename is not a prerequisite for this PR, so I'd rather keep the PR contained. Thanks!

lersek replied:
... Yet another alternative is (of course) just to append patch v2 nr.17, implementing the rename you suggest. That seems doable, to me; would it work for you?

rminnich replied:
fullCopy copies ...

jensdrenhaus replied:
Not too important. I leave it up to you.

rminnich commented on cmds/core/netcat/listen.go:
```diff
@@ -110,31 +109,102 @@ func (c *cmd) setupListener(network, address string) (net.Listener, error) {
 
-func NewConnections() *Connections {
-	return &Connections{
+func NewConnections(capacity uint32) *Connections {
```
// NewConnections ..

lersek replied:
This is from preexistent code (I've only modified the parameter list), so I didn't want to add a comment.

From the other discussion with Jens, I've [ended up](https://github.com/u-root/u-root/pull/3337#discussion_r2024481272) renaming (lower-casing) this stuff in v2, in a new patch (nr.17). That, combined with your [note](https://github.com/u-root/u-root/pull/3337#discussion_r2023568797) that the linter might enforce comments regardless, might cause me to add a brand new comment here. If the linter is happy with no comment on `newConnections` though, then I wouldn't like to add one -- I wouldn't like to get into adding comments to all prior functions that had not been documented before. This series is already sprawling.

(I'm resolving this thread now, because CI will force v2 through linting in any case.)

rminnich commented on cmds/core/netcat/listen.go:
```diff
@@ -110,31 +109,102 @@ func (c *cmd) setupListener(network, address string) (net.Listener, error) {
+	isEmpty     *sync.Cond
 }
 
-func NewConnections() *Connections {
```
I don't understand the need for a capacity -- surely we can let the OS enforce that. 

But the use of sync.Cond etc. here looks like an opportunity to have serious races. 

A common way to do bounded allocators of this type is with a chan that has a fixed size. Allocation is performed by writing to it; if It is at limit, the write will block. The Delete is a free. But maybe chan is not a good fit to what you want. But what does the map provide? 

I'm uncertain about this design.

RiSKeD replied:
I am unsure myself, but wouldn't a [sync.WaitGroup](https://cs.opensource.google/go/go/+/refs/tags/go1.24.2:src/sync/waitgroup.go) be the right fit? 
-- increment it on adding a connection, decrement it on deletion, and add its wait-condition for the drain function.

lersek replied:
@rminnich

> I don't understand the need for a capacity -- surely we can let the OS enforce that.

We could, but `netcat` has an option that requires this capacity concept:

~~~
  -m uint
        Maximum <n> simultaneous connections (shorthand) (default 100)
  -max-conns uint
        Maximum <n> simultaneous connections (default 100)
~~~

As I write in the commit message of patch v1 nr.16, the central problem with the prior listen-mode implementation is that it conflates *max number of simultaneous connections* with *max number of Accept() attempts*. The former is something that the `netcat` command line exposes, and the latter is only useful for unit-testing keep-open mode / broker mode. And, they are mostly independent of each other. (I'm saying "mostly" because you obviously cannot have more concurrent connections than you're willing to call `Accept`.) I needed to disentangle both concepts in the code, which basically requires maintaining and enforcing said "capacity".

> But the use of sync.Cond etc. here looks like an opportunity to have serious races.
> 
> A common way to do bounded allocators of this type is with a chan that has a fixed size. Allocation is performed by writing to it; if It is at limit, the write will block. The Delete is a free. But maybe chan is not a good fit to what you want. But what does the map provide?
> 
> I'm uncertain about this design.

There are four layers to the design of the `Connections` type (renamed to `connections` in v2):
1. The map is an associative array for storing live connections. It is associative because connections come and go, and while adding a new connection could be just "append", deleting a connection that just completed must be random-access. So using a map (in preexistent code) is a proper choice, IMO.

   (Well, in theory, a list would do as well, if connections weren't identified numerically -- the numeric connection ID gets assigned in the order connections are accepted. If each connection held a pointer to its location (= node) in this list, it could remove itself O(1) as well, but managing lists like this is probably too low level for Go code. Either way, the existent map was just fine for the need.)
  
2. The map is protected with a mutex, because the map is accessed concurrently (still in the preexistent code) by multiple goroutines. (a) The main goroutine, running the `Accept` loop, extends the map (with the `Add` method). (b) Each connection-specific goroutine deletes the connection from the map when the connection completes. (c) When we are in broker mode, every connection-specific goroutine broadcasts the data it receives from its connection to all other connections. For this, said goroutine iterates over the entire map. The pre-patch code handles all this correctly; whichever goroutine needs access, it acquires the mutex, for the full length of the operation in question. More performant solutions (lockless data structures) may exist in theory, but the connection map is not a highly contended resource, so a mutex is just fine, IMO.

3. The third layer is the first of _those that I'm adding_. This layer fits fully naturally into the prior design/code (i.e., the above two layers).

   Pre-patch, the map can grow boundlessly. That's not what the `--max-conns` option dictates, though. For `--max-conns`, we need to extend the map with (a) usage tracking, and (b) the ability to block the main loop, in `Add`, from accepting further connections, when the map is "full".

   There are multiple ways to do this. Conceptually, we need a counting semaphore, initialized to full capacity, then perform a DOWN op in `Add`, and perform an UP op in `Delete`. As you say, a "chan that has a fixed size" would work just fine. However, given that we _already_ have a mutex-protected data structure -- which is _correct_ in itself, see above --, the *natural* choice is to introduce a usage counter (also protected by the mutex), and a condition variable.

   A condition variable:
   - is associated with a mutex-protected data structure, and a predicate P expressed on the fields of the protected data structure,
   - lets users of the data structure go to sleep if predicate P does not hold,
   - wakes sleeping user(s) of the data structure whenever predicate P *transitions* from `false` to `true`.

   In our case, predicate P is, "the map is not full" (that is, `used != capacity`). This is the `isAvailable` condvar in patch v1 nr.16.

   The [Go documentation](https://pkg.go.dev/sync#Cond) describes condvars exactly as I expected (I had extensively worked with POSIX threads, mutexes and condvars in the past), so I'm comfortable using them in Go.

   Again, using a counting semaphore, or a fixed-size channel would be fine, functionally speaking. However, those primitives use their *own* locking and condition variables internally! And so it is conceptually wasteful to have a mutex-protected map for our connections, *plus* a mutex-protected, and (inevitably) condvar-using, extra data structure, such as a semaphore, or a fixed-size chan. It's cleaner to add a condvar to our already existent mutex-protected data structure, and express the predicate that we care about, on that same data structure.

4. (This is the second layer of those that I'm adding.) In particular, the nice thing about condvars is that you can associate multiple predicates with the same one data structure / same one mutex -- the invariants of the structure do _not_ change; the idea is _only_ that you allow different users to block until different predicates become true.

   In our case, the second predicate happens to originate from the unit testing requirement that the `Accept` loop in the main goroutine be bounded, in keep-open or broker mode. (In production, the `Accept` loop is never supposed to terminate, in keep-open or broker mode.) Again, let me emphasize that this limit is *totally different* from the simultaneous connection limit.

   Now think of what happens when we hit this `Accept` iteration limit -- the main goroutine is supposed to exit the `Accept` loop, but there may still be a number of sub-goroutines running (each handling a connection accepted earlier). Before we can let the main goroutine return, we need to wait for all those connections to complete. But that's just another predicate for our mutex-protected data structure: the main goroutine needs to block, in the end, until the usage counter falls back down to zero. (This is what `Drain` does.) So, introduce another condvar for this predicate: `isEmpty`; it holds true when `used == 0`.

   You *could* implement this 4th layer with a Go WaitGroup. But note how the *hidden* synchronization elements would proliferate:
   - as a basis, we'd have a mutex-protected map, from before,
   - we'd have a fixed-size channel for enforcing `--max-conns`, and this channel would come with its *own* internal counter, mutex and condvar,
   - we'd have a WaitGroup for letting connections drain (for unit testing's sake), which *too* would bring its *own* internal counter, mutex and condition variable.

   The problem is not that this solution is "slow" -- again, the connection map is not a contended data structure. The problem is that this model is *conceptually wasteful*. Once you've used condition variables, you can't quite miss that the "fixed-size chan + waitgroup" approach uses two mutexen, and one counter, *more* than what's required:
   - "fixed-size chan + waitgroup" uses thee mutexes (map, channel, waitgroup), two counters (channel, waitgroup), and two condvars (channel, waitgroup)
   - the explicit condvar code in patch v1 nr.16 uses one mutex (for protecting the map), one counter (for tracking the number of simultaneous connections), and two condvars (two predicates expressed on the same one counter).

   The conceptual waste in the former option is therefore two mutexes plus one counter.

lersek replied:
@RiSKeD 

> I am unsure myself, but wouldn't a [sync.WaitGroup](https://cs.opensource.google/go/go/+/refs/tags/go1.24.2:src/sync/waitgroup.go) be the right fit? -- increment it on adding a connection, decrement it on deletion, and add its wait-condition for the drain function.

This would indeed cover `drain`, but it would not enforce the limit on the number of simultaneous connections.

My [point](https://github.com/u-root/u-root/pull/3337#discussion_r2024655351) is that you don't need two distinct synchronization abstractions like fixed-size chan *plus* WaitGroup. These two distinct abstractions maintain the *current connection count* doubly, which is a mental waste. It suffices to maintain the *current connection count* concept only once, protect it together with the rest of the connection map, under the same (preexistent) mutex, and express the two (indeed different) predicates that we care about using two condition variables, formulated over the same one mutex and same one connection counter.

lersek replied:
Oh, I also wanted to add: if I were writing this logic from scratch, I would likely implement it in small stages, using small patches. Unfortunately, I can't do it here: bisectability of the git history does not allow me to regress the existent listen mode impl. even temporarily, so I need to do the switch-over in one go. *That* is certainly quite unfortunate.

jensdrenhaus replied:
I have not seen `sync.Cond` a lot yet and on first look I also thought that a buffered channel (even without a sync.WaitGroup`) could do the job. Even `sync.Cond`s docu suggests that: 
```
...
For many simple use cases, users will be better off using channels than a Cond (Broadcast corresponds to closing a channel, and Signal corresponds to sending on a channel).
...
```

But after all this will not cover all needed aspects here as @lersek described in detail above. That being said, the concept of `sync.Cond` is used as intended here. I think we are good to go with this implementation.

rminnich commented on pkg/netcat/util.go:
```diff
@@ -60,3 +65,115 @@ func (sw *ConcurrentWriter) Write(p []byte) (n int, err error) {
+	return os.Stdout.Write(b)
+}
+
+func (swc *StdoutWriteCloser) Close() error {
```
The commend can be as simple as Close implements io.Close or whatever, but the linter will want a comment.

jensdrenhaus replied:
https://golangci-lint.run/welcome/install/#local-installation
and then `golangci-lint run` in u-root root directory

binjip978 replied:
We don't this particular linter, but indeed golangcilint is failing, you can run it locally:
```
go install github.com/golangci/golangci-lint/cmd/golangci-lint@v1.64.5
GOOS=linux GOOS=amd64 golangci-lint run ./... # from u-root directory
```

lersek replied:
@jensdrenhaus @binjip978 thanks, I've managed to install it!

Added StdoutWriteCloser comments in v2.

lersek commented on cmds/core/netcat/listen.go:
```diff
@@ -110,31 +109,102 @@ func (c *cmd) setupListener(network, address string) (net.Listener, error) {
 	}
 }
 
+// Accept a connection from "listener" such that the peer address is permitted
```
Yes, thanks! Done in v2, per Jens' [earlier comment](https://github.com/u-root/u-root/pull/3337#issuecomment-2769768328).

This documentation style is very different from what I'm used to in C (and in commit messages), where generally imperative mood is preferred (IOW my comments on new functions were already consistent in v1, at least that was my indent; it's just that they were consistent with a style that's not proper for Go. :smile: )

Lesson learned, thanks.

lersek commented on cmds/core/netcat/listen.go:
```diff
@@ -158,106 +228,170 @@ func (c *Connections) Broadcast(output io.Writer, senderID uint32, message strin
-		wg                 sync.WaitGroup
-		once               sync.Once
-	)
+// Main transfer routine for the listener if neither keep-open nor broker mode
```
Indeed; after Jens gave me the same comment, I went through the [entire patch set](https://github.com/u-root/u-root/pull/3337#issuecomment-2769768328), and fixed up all the comments.

lersek commented on cmds/core/netcat/listen.go:
```diff
@@ -110,31 +109,102 @@ func (c *cmd) setupListener(network, address string) (net.Listener, error) {
 
+// Add a new connection. If the number of concurrent connections has reached
+// the maximum (i.e., capacity has been exhausted), block until another
+// goroutine calls Delete().
```
Yes, [done](https://github.com/u-root/u-root/pull/3337#issuecomment-2769768328) in v2. Thanks.

lersek commented:
Updates in v2:
- patch 1 (`pkg/netcat: add StdoutWriteCloser type`):
  - drop useless constructor `NewStdoutWriteCloser` [Jens]
  - add function comments [Ron]
  - add stub `Close` impl for non-UNIX, using build constraints and separate source files [Jens, Ron]
  - add POSIX references to the type docs
- patch 2 (`pkg/netcat: add TeeWriteCloser type`):
  - update comments on `TeeWriteCloser`, `TeeWriteCloser.Write`, and `TeeWriteCloser.Close` to follow Go documentation conventions [Jens, Ron]
- patch 4 (`pkg/netcat: satisfy the io.WriteCloser interface with OutputOptions`):
  - add comment to `OutputOptions.Close` [Jens]
- patch 5 (`pkg/netcat: add CloseWrite() helper function`):
  - update comment on `CloseWrite()` to follow Go documentation conventions [Jens, Ron]
- patch 7 (`cmds/core/netcat: "close" stdout properly`):
  - open-code useless constructor `netcat.NewStdoutWriteCloser` [Jens]
- patch 16 (`cmds/core/netcat: rewrite listenForConnections()`):
  - update comments on `acceptAllowed`, `fullCopy`, `Connections.Add`, `acceptSingle`, and `acceptForever`, to follow Go documentation conventions [Jens, Ron]
  - rename `testId` to `testID`, `connId` to `connID` [lint]
  - rename `myConnId` to `myConnID` (just in case; not flagged by lint)
  - `acceptAllowed`: adhere to *do not name result parameters just because it enables you to use naked returns*, from https://google.github.io/styleguide/go/decisions#named-result-parameters [Jens]
- patch 17 (`cmds/core/netcat: unexport the Connections struct`):
  - new patch [Jens]

In addition to the [previously mentioned](https://github.com/u-root/u-root/pull/3337#issue-2961055615) per-patch checks, I've now also linted the u-root tree at every patch in the series (I'll add this fact to the PR description in a minue).

Comment on the workflow: Github requires us to force-push the same branch for updating a PR. This is an awful practice, because it makes it very difficult for reviewers to perform a *patch-wise incremental review*. To add insult to injury, after a force-push, Github also does not allow reviewers to fetch previous versions of the branch, in case they wanted to perform an incremental review *locally*. Therefore, I have now pushed both the previous (v1) and the current (v2) version of this branch (`netcat-eof-prop-3329`), to my u-root fork, under separate names `netcat-eof-prop-3329-v1` and `netcat-eof-prop-3329-v2`. The commit hashes are (just for clarity):
- current version (`netcat-eof-prop-3329`, `netcat-eof-prop-3329-v2`): `857b8c1caadc`
- previous version (`netcat-eof-prop-3329-v1`): `107a52de7be3`

If you'd like to do an incremental I review, I recommend fetching both `netcat-eof-prop-3329-v1` and `netcat-eof-prop-3329-v2`, and then running (in your local clone):
~~~
git range-diff --color main netcat-eof-prop-3329-v1 netcat-eof-prop-3329-v2 | less
~~~

lersek commented:
Regarding v2 CI failures:
- the codecov whine about `exp` can only be wrong; I didn't touch `exp`
- most other codecov complaints will be covered by the upcoming netcat integration test that I mention in the PR desc. (As the gosh bug mentioned there blocks the loopback test, I've added several other tests, since.)
- [This one](https://github.com/u-root/u-root/actions/runs/14221820763/job/39851534913?pr=3337) seems like a valid complaint. I expected it to be resolved by restricting the dup2 syscall to UNIX geese, but that change seems insufficient for "collecting and putting dependencies in place". I'll look into that for v3.
- [This one](https://github.com/u-root/u-root/actions/runs/14221820756/job/39851534954?pr=3337) also persists, and I don't understand it. It complains about dup2 (i.e., TestGoTest failing to *build* because of `undefined: syscall.Dup2`) -- but dup2 is certainly available on arm64 Linux! I've got no idea what's happening here. Please help! Thanks.

Ahh, wait, the lasts two complaints are *both* for ARM64. Something is fishy with cross-compiling to ARM64, then!

lersek commented:
What's awful about this is that POSIX-like OSes that are not even Linux, but *do* implement dup2 as a syscall, will know nothing about dup3. Therefore I can't just replace dup2 with dup3, because it could break other (non-Linux) POSIX-like OSes. I need to make an exception form ARM64 linux specifically. Awful.

lersek commented:
v3 (also `netcat-eof-prop-3329-v3` in my fork):
- patch 1 (`pkg/netcat: add StdoutWriteCloser type`):
  - special-case `dup2` for ARM64 Linux
  - elaborate more on file description reference counting in the leading comment of the `StdoutWriteCloser` type [Jens]

lersek commented:
v4 (also `netcat-eof-prop-3329-v4` in my fork):
- patch 1 (`pkg/netcat: add StdoutWriteCloser type`):
  - also special-case `dup2` for RISC-V64 Linux

lersek commented:
Hallelujah, v4 passed CI.

(Not counting the code coverage checks, which we should ignore, as the upcoming netcat integration test will cover those new code paths.)

jensdrenhaus reviewed (approved):
Great job!

lersek commented:
Thank you all for the reviews and/or approvals!
